Add app-owned buffers user documentation (#4872)
microsoft/msquic: docs/Streams.md

The QUIC protocol allows a maximum number of streams equal to 2 ^ 62. As there are 4 unique stream types, the maximum number of streams is 2 ^ 60, per stream type. No app would likely need to have this many streams open at any point.

For this reason, each app controls the number of streams that the peer is allowed to open. The concept is similar to flow control of the actual data on a stream. The app tells the peer how many streams it's willing to accept at any point. Instead of a buffer size, it's a stream count.

> G

Author: ? Remove this.

For asynchronous receives, the app indicates how much of the data it accepted via the `BufferLength` parameter passed into the [StreamReceiveComplete](api/StreamReceiveComplete.md) API.

Any value less than or equal to the initial **TotalBufferLength** value is allowed, including zero.

Whenever a receive isn't fully accepted by the app, additional receive events are immediately disabled. The app is assumed to be at capacity and not able to consume more until further indication. To re-enable receive callbacks, the app must call [StreamReceiveSetEnabled](api/StreamReceiveSetEnabled.md).

There are cases where an app may want to partially accept the current data, but still immediately get a callback with the rest of the data. To do this (only works in the synchronous flow) the app must return `QUIC_STATUS_CONTINUE`.

> ## Receive modes

Reviewer: ```suggestion
## Receive Modes
```

Options can be used to alter MsQuic default receive notification behavior:

> ### Multi-receive mode

Reviewer: ```suggestion
### Multi-Receive Mode
```

Multi-receive mode is a connection wide option allowing multiple receive notification to be pending simultaneously.

> It is enabled by setting [`StreamMultiReceiveEnabled`](./Settings.md) in connection parameters.

Reviewer: copy/paste wrong link

Author: ```suggestion
It is enabled by setting [`StreamMultiReceiveEnabled`](./api/QUIC_SETTINGS.md) in connection parameters.
```

For streams created when the connection is in Multi-mode receive, MsQuic can keep indicating `QUIC_STREAM_EVENT_RECEIVE` before the application completes the previous one.
This means that the application must be able to handle a new `QUIC_STREAM_EVENT_RECEIVE` even if it returned `QUIC_STATUS_PENDING` previously and has not called [`StreamReceiveComplete`](api/StreamReceiveComplete.md) yet.

> MsQuic will also keep indicating receive notifications when the application accept the data partially. The bytes that have not been accepted by the application won't be indicated again: the application must call [StreamReceiveComplete](api/StreamReceiveComplete.md) in the future to accept them.

Reviewer: ```suggestion
MsQuic will also keep indicating receive notifications when the application accepts the data partially. The bytes that have not been accepted by the application won't be indicated again: the application must call [StreamReceiveComplete](api/StreamReceiveComplete.md) in the future to accept them.
```

> To handle multi-receive mode properly, the application must keep of the total number of byte received on the stream (the sum of all `TotalBufferLength`).

Reviewer: ```suggestion
To handle multi-receive mode properly, the application must keep track of the total number of byte received on the stream (the sum of all `TotalBufferLength`).
```

> The number of calls to [`StreamReceiveComplete`](api/StreamReceiveComplete.md) does not need to be equal to the number of receive notification, but the total number of bytes received must eventually be equal to the total number of bytes accepted.

Reviewer: ```suggestion
The number of calls to [`StreamReceiveComplete`](api/StreamReceiveComplete.md) does not need to be equal to the number of receive notification, but the total number of bytes completed must eventually be equal to the total number of bytes received.
```

Multi-receive mode manages its internal receive buffer differently and is more efficient for continuous receiving with asynchronous processing.

> ### App-owned buffer mode

Reviewer: ```suggestion
### App-Owned Buffer Mode
```

App-owned buffer mode is a per-stream option allowing the application to provide its own receive memory buffers.
Enabling app-owned mode is done differently depending on whether the stream is created locally or from the peer and is discussed below.

When in app-owned mode, the application can call [`StreamProvideReceiveBuffers`](./????.md) to provide a list of memory buffers to MsQuic.

> [`StreamProvideReceiveBuffers`](./????.md) can be called at any time on a valid stream in app-owned mode, potentially inline from a notification handler.

Reviewer: TODO: Add at least stubs for these `????` files.

If called several times, the buffer provided through subsequent calls are added to the list.

> MsQuic will fill the provided buffer with received data, in the order they have been provided. Receive notification will be emitted as normal,
> indicating a list of `QUIC_BUFFER` pointing to the application provided buffers.
> Note that up to the number of buffers the application provided can be indicated at once, and that only part of a buffer can be indicated.
> There is no guarantees the `QUIC_BUFFER`s indicated in a receive notification will match the ones the application provided.

Reviewer: ```suggestion
MsQuic will fill the provided buffer(s) with received data, in the order they have been provided. Receive notification will be emitted as normal,
indicating a list of `QUIC_BUFFER` pointing to the application provided buffer(s).
Note that up to the number of buffers the application provided can be indicated at once, and that only part of a buffer can be indicated.
There is no guarantee the `QUIC_BUFFER`s indicated in a receive notification will match the ones the application provided.
```

The application is responsible for tracking the amount of data received and when a buffer it provided has been fully used.

> The application regains full ownership of a buffer after it get a receive notification for all bytes in the buffer and accept them by calling [`StreamMultiReceiveEnabled`](./Settings.md).

Reviewer: ```suggestion
The application regains full ownership of a buffer after it get a receive notification for all bytes in the buffer and accepts them by calling [`StreamReceiveComplete`](./Settings.md).
```
It's receive complete, not receive enabled? Also, please fix up all these links that point to `Settings.md`

Author: Yes, that looks like a copy-past gone wrong.

If the application accepts all the buffer's bytes **inline** from the receive notification, by returning `QUIC_STATUS_SUCCESS` and setting `TotalBufferLength` appropriately,
it can free or reuse the buffer while in the notification handler.

For an application, providing receive buffers can improve performances by saving a copy: MsQuic places data directly in its final destination.
However, it comes with a large complexity overhead for the application, both in term of memory management and in term of flow control:

> an application providing to much or too little buffer space could negatively impact performances.

Reviewer: ```suggestion
an application providing too much or too little buffer space could negatively impact performances.
```

Because of this, app-owned mode should be considered an advanced feature and used with caution.

> **Note**: As of now, app-owned buffer mode is not compatible with multi-receive mode. If multi-receive mode is enabled for the connection and app-owned mode is enabled on a stream, that specific stream will behave as if multi-receive mode was disabled. This may change in the future.

Reviewer: ```suggestion
> **Note**: As of now, app-owned buffer mode is not compatible with multi-receive mode. If multi-receive mode is enabled for the connection and app-owned mode is enabled on a stream, that specific stream will behave as if multi-receive mode was disabled. This may change in the future.
```

> #### For a stream initiated by the application

Reviewer: ```suggestion
#### Locally Initiated Streams
```

To use app-owned buffers on a locally created stream, the flag `QUIC_STREAM_OPEN_FLAG_APP_OWNED_BUFFERS` must be provided to the [`StreamOpen`](./api/StreamOpen.md).

> Before starting the stream with [`StreamStart`](./api/StreamStart.md), the application should call [`StreamProvideReceiveBuffers`](./api/StreamProvideReceiveBuffers.md) to provide some initial buffers.

Reviewer: "should call"... this is not a _must call_?

Author: Nothing will prevent you from starting a stream in App-owned mode without providing buffers, and everything will be fine as long as you don't receive any data on it (or provide buffers before you do receive data). So technically, it's a "should", but it would be hard to find a reasonable scenario where it makes sense not to provide buffers before starting the connection.

I can make this a must if you think it makes it less confusing.

> **Note**: This is only relevant for a bidirectional stream, since a locally created unidirectional stream cannot receive data.

Reviewer: ```suggestion
> **Note**: This is only relevant for a bidirectional stream, since a locally created unidirectional stream cannot receive data.
```

> #### For a stream initiated by a peer

Reviewer: ```suggestion
#### Peer Initiated Streams
```

> To use app-owned buffers on a peer initiated stream, the application should call [`StreamProvideReceiveBuffers`](./api/StreamProvideReceiveBuffers.md) inline when handling the `QUIC_CONNECTION_EVENT_PEER_STREAM_STARTED` notification.

Reviewer: Again, "should" or "must"?

Author: This one is a must, fixed it.

When called inline while handling `QUIC_CONNECTION_EVENT_PEER_STREAM_STARTED`, [`StreamProvideReceiveBuffers`](./api/StreamProvideReceiveBuffers.md) enables app-owned buffers and provides some initial buffers. This is the only situation where it is allowed to call [`StreamProvideReceiveBuffers`](./api/StreamProvideReceiveBuffers.md) on a stream that is not already in app-owned buffers mode. After this initial call, [`StreamProvideReceiveBuffers`](./api/StreamProvideReceiveBuffers.md) can be called at any time to provide more buffer space, until the stream is closed.

> #### Initial buffer space

Reviewer: ```suggestion
#### Initial Buffer Space
```

> As part of the connection establishment, Quic exchanges initial stream flow control limit as part of the transport parameters, defining the amount of data the each peer will be allowed to send on a newly created stream. An application can define these limits through `StreamRecvWindowBidiLocalDefault`, `StreamRecvWindowBidiRemoteDefault` and `StreamRecvWindowUnidiDefault` in [`QUIC_SETTINGS`](./api/QUIC_SETTINGS.md).

Reviewer: ```suggestion
As part of the connection establishment, QUIC exchanges initial stream flow control limit as part of the transport parameters, defining the amount of data that each peer will be allowed to send on a newly created stream. An application can define these limits through `StreamRecvWindowBidiLocalDefault`, `StreamRecvWindowBidiRemoteDefault` and `StreamRecvWindowUnidiDefault` in [`QUIC_SETTINGS`](./api/QUIC_SETTINGS.md).
```
